README.md
# lux-autohost

## What is it?

lux-autohost provides two Action Creator API abstractions (for logging and metrics). These APIs provide developer-friendly (& expressive) methods that wrap the publishing of lux Action messages:

* Logging Action Creator API
	* `error( data )`
	* `warn( data )`
	* `info( data )`
	* `debug( data )`
* Metrics Action Creator API
	* `meter( key, value [, unit, customData ] )`
	* `timer( key, value [, customData ] )` (timer units are in ms)

In addition to these methods, lux-autohost transparently adds metrics-capture for any lux Action executed in the client. In other words, lux-autohost automatically starts count and timer metrics for every action (so you don't have to worry about that), and it provides the Action Creator APIs in case you have other custom metrics/logging you need to capture.

Logging and metric payloads will be batched and transmitted at the interval you specify (see below).

## Usage

### Configuring the interval(s)

Below is an example of configuring the intervals at which logging & metrics will be transmitted to the autohost resource endpoints. In this example, metrics-related messages will be transmitted every 30 seconds, or every 500 messages (whichever is reached first), and logging-related messages will be transmitted every 5 seconds, or every 25 messages.

```javascript
define([
	"lux-autohost"
], function( luxah ) {
	luxah.config({
		metrics: {
			timeout: 30000, // every 30 seconds...OR
			messages: 500   // every 500 messages
		},
		logging: {
			timeout: 5000, // every 5 seconds...OR
			messages: 25   // every 25 messages
		}
	});
});
```

### Configuring include/exclude

By default, lux-autohost will wire up metrics on *every* action. You may have actions that you don't care about monitoring (for example: an action that fires any time a text input updates as a user types). You can choose to add actions to an "exclude" list, OR you can switch lux-autohost into "include" mode, and only the actions specified will be monitored.

```javascript
define([
	"lux-autohost"
], function( luxah ) {
	luxah.config({
		filter: {
			include: false, // default value is false - we exclude by default
			actions: [ "someAction", "anotherAction" ]
		},
		// interval configuration example from above snippet
		// showing you can set all this config in one call
		metrics: {
			timeout: 30000, // every 30 seconds...OR
			messages: 500   // every 500 messages
		},
		logging: {
			timeout: 5000, // every 5 seconds...OR
			messages: 25   // every 25 messages
		}
	});
});
```

### Using the Action Creator APIs

#### Logging

Here's an example of a React component pulling in the `error`, `warn`, `info` and `debug` Action Creator API calls that are provided for logging.

```javascript
define( [
	"react",
	"lux.js",
], function( React, lux ) {
	var Component = React.createClass( {

		mixins: [ lux.reactMixin.actionCreator ],

		getActions: [ "error", "warn", "info", "debug" ],

		componentDidMount: function() {
			this.debug( "Yay! My component mounted." );
		},

		componentWillUnmount: function() {
			this.warn( "Sad times, my component is going to unmount." );
		},

		handleFatalError: function( data ) {
			this.error( data );
		},

		// more methods, etc.
	} );

	return Component;
} );

```

#### Metrics

Here's an example of a React component pulling in the `meter` and `timer` Action Creator API calls that are provided for metrics.

```javascript
define( [
	"react",
	"lux.js",
], function( React, lux ) {
	var Component = React.createClass( {

		mixins: [ lux.reactMixin.actionCreator ],

		getActions: [ "meter", "timer" ],

		componentDidMount: function() {
			this.meter("some.arbitrary.key", 1);
		},

		componentWillUnmount: function() {
			this.meter( "gauge.the.thing", -20 );
		},

		render: function() {
			var timer = this.timer( "pointless.render.timer" );
			// do render-y things
			timer.record();
		},

		// more methods, etc.
	} );

	return Component;
} );

```

### How Logging and Metrics Make it to autohost

The underlying batch manager in lux-autohost is an Action Listener for actions matching the method names provided by the logging and metrics Action Creator APIs (in other words, it listens for `timer`, `meter`, `error`, `warn`, `info` and `debug` action messages). Logging and metrics action messages are queued until the interval threshold has been reached, at which point lux-autohost will publish a `sendLogBatch` or `sendMetricsBatch` action message. Your API wrapper will need to handle `sendLogBatch` and/or `sendMetricsBatch` actions, transmitting them to your autohost endpoints however you prefer (we use [halon](https://github.com/LeanKit-Labs/halon) to do this). For example, our API wrapper might look like this:

```javascript
define( [
	"lux.js",
	"halon",
	"jquery"
], function( lux, halon, $) {

	var lk = halon( {
			root: window.location.origin + "/api",
			knownOptions: {
				logging: [ "upload" ],
				metrics: [ "upload" ]
			},
			adapter: halon.jQueryAdapter( $ ),
			version: 1
		} );

	lk.connect();

	return lux.actionCreatorListener( {
			namespace: "muhapi",
			handlers: {
				// Other API wrapper methods....
				sendLogBatch: function( batch ) {
					lk.logging.upload( data ).catch( function( err ) {
						if ( DEBUG ) {
							console.log( "Unable to reach logging endpoint: ", err );
						}
					} );
				},
				sendMetricsBatch: function( batch ) {
					lk.metrics.upload( data ).catch( function( err ) {
						if ( DEBUG ) {
							console.log( "Unable to reach metrics endpoint: ", err );
						}
					} );
				}
			}
		} );
} );

```


CHANGELOG.md
## 1.x

### 1.0.0

* Updated to lux 2.0.0-1



spec/batchManager.spec.js
describe( "batchManager.js", function() {
	var clock;
	before( function() {
		clock = sinon.useFakeTimers();
	} );
	after( function() {
		clock.restore();
	} );
	describe( "when batching log entries", function() {
		var listener, creator;
		function sendLogMessage() {
			creator.info( "And now for some important info...." );
		}
		beforeEach( function() {
			initLuxAh();
			listener = global.lux.actionListener( {
				handlers: {
					sendLogBatch: sinon.stub(),
					sendLogEntry: sinon.stub()
				}
			} );
			creator = lux.actionCreator( { getActions: [ "error", "warn", "info", "debug" ] } );
			global.lah.loggingClient.queue = [];
		} );
		afterEach( function() {
			listener.luxCleanup();
		} );
		it( "should queue the entry if the threshold hasn't been reached", function() {
			sendLogMessage();
			listener.handlers.sendLogBatch.should.not.have.been.called;
			listener.handlers.sendLogEntry.should.have.been.calledOnce;
		} );
		it( "should publish a sendLogBatch action if the timeout threshold has been reached", function() {
			sendLogMessage();
			listener.handlers.sendLogEntry.should.have.been.calledOnce;
			clock.tick( 10000 );
			listener.handlers.sendLogBatch.should.have.been.called;
		} );
		it( "should publish a sendLogBatch action if the message count threshold has been reached", function() {
			var i = 25;
			while ( i ) {
				sendLogMessage();
				i--;
			}
			listener.handlers.sendLogEntry.callCount.should.equal( 25 );
			listener.handlers.sendLogBatch.should.have.been.called;
		} );
		it( "should empty the log queue once the batch has been sent", function() {
			var i = 25;
			while ( i ) {
				sendLogMessage();
				i--;
			}
			global.lah.loggingClient.queue.length.should.equal( 0 );
		} );
	} );
	describe( "when batching metrics entries", function() {
		var listener, creator;
		function sendMetricsMessages() {
			creator.meter( "some.random.key", 42 );
			creator.timer( "time.keeps.on.tickin", 42000 );
		}
		before( function() {
			initLuxAh();
		} );
		beforeEach( function() {
			listener = global.lux.actionListener( {
				handlers: {
					sendMetricsBatch: sinon.stub(),
					sendMetricsEntry: sinon.stub()
				}
			} );
			creator = lux.actionCreator( { getActions: [ "meter", "timer" ] } );
			global.lah.metricsClient.queue = [];
		} );
		afterEach( function() {
			listener.luxCleanup();
		} );

		it( "should queue the entry if the threshold hasn't been reached", function() {
			sendMetricsMessages();
			listener.handlers.sendMetricsBatch.should.not.have.been.called;
			listener.handlers.sendMetricsEntry.should.have.been.calledTwice;
		} );
		it( "should publish a sendMetricsBatch action if the timeout threshold has been reached", function() {
			sendMetricsMessages();
			listener.handlers.sendMetricsEntry.should.have.been.calledTwice;
			clock.tick( 35000 );
			listener.handlers.sendMetricsBatch.should.have.been.called;
		} );
		it( "should publish a sendMetricsBatch action if the message count threshold has been reached", function() {
			for ( var i = 0; i < 250; i++ ) {
				sendMetricsMessages();
			}
			listener.handlers.sendMetricsEntry.callCount.should.be.within( 500, 502 );
			listener.handlers.sendMetricsBatch.should.have.been.called;
		} );
		it( "should empty the metrics queue once the batch has been sent", function() {
			for ( var i = 0; i < 250; i++ ) {
				sendMetricsMessages();
			}
			global.lah.metricsClient.queue.length.should.equal( 0 );
		} );
	} );
} );


spec/configuration.spec.js
describe( "Configuration", function() {
	beforeEach( function() {
		initLuxAh();
	} );

	it( "should start with defaults", function() {
		global.lah.config().should.eql( {
			actionChannel: postal.channel( "lux.action" ),
			filter: {
				include: false,
				actions: []
			},
			metrics: {
				timeout: 30000,
				messages: 500
			},
			logging: {
				timeout: 5000,
				messages: 25
			}
		} );
	} );

	it( "should allow the action channel to be set", function() {
		global.lah.config( { actionChannel: postal.channel( "customChannel" ) } );
		global.lah.config().actionChannel.should.eql( postal.channel( "customChannel" ) );
	} );

	it( "should allow the metrics options to be set", function() {
		global.lah.config( {
			metrics: {
				messages: 4242,
				timeout: 42000
			}
		} );
		global.lah.config().metrics.should.eql( {
			messages: 4242,
			timeout: 42000
		} );
	} );

	it( "should allow the logging options to be set", function() {
		global.lah.config( {
			logging: {
				messages: 4242
			}
		} );
		global.lah.config().logging.should.eql( {
			messages: 4242,
			timeout: 5000
		} );
	} );
} );


spec/helpers/node-setup.js
var jsdom = require( "jsdom" );
global.document = jsdom.jsdom( "<html><body></body></html>" );
global.window = document.defaultView;
global.BROWSER = false;
jsdom.changeURL( global.window, "http://localhost/test?queryString=1" );

require( "babel-polyfill" );
var chai = require( "chai" );
chai.use( require( "sinon-chai" ) );
global.should = chai.should();
global.proxyquire = require( "proxyquire" ).noPreserveCache().noCallThru();

global.postal = require( "postal" );
global.machina = require( "machina" );
global._ = require( "lodash" );
global.sinon = require( "sinon" );
global.moment = require( "moment" );
global.lux = undefined;
global.lah = undefined;

require( "babel-register" )( {
	only: /spec/
} );

global.initLuxAh = function() {
	if ( global.lah ) {
		global.lah.batchListener.luxCleanup();
	}

	if ( global.lux ) {
		global.lux.dispatcher.dispose();
	}

	global.lux = global.proxyquire( "lux.js", {} );
	global.lah = global.proxyquire( "../../lib/lux-autohost.js", {
		"lux.js": global.lux
	} );
};


gulpfile.js
var gulp = require( "gulp" );
var gutil = require( "gulp-util" );
var sourcemaps = require( "gulp-sourcemaps" );
var rename = require( "gulp-rename" );
var header = require( "gulp-header" );
var imports = require( "gulp-imports" );
var pkg = require( "./package.json" );
var hintNot = require( "gulp-hint-not" );
var uglify = require( "gulp-uglify" );
var babel = require( "gulp-babel" );
var jscs = require( "gulp-jscs" );
var gulpChanged = require( "gulp-changed" );

var banner = [ "/**",
	" * <%= pkg.name %> - <%= pkg.description %>",
	" * Author: <%= pkg.author %>",
	" * Version: v<%= pkg.version %>",
	" * Url: <%= pkg.homepage %>",
	" * License(s): <% pkg.licenses.forEach(function( license, idx ){ %><%= license.type %> Copyright (c) <%= ( new Date() ).getFullYear() %> LeanKit<% if(idx !== pkg.licenses.length-1) { %>, <% } %><% }); %>",
	" */",
"" ].join( "\n" );

gulp.task( "build:es6", function() {
	return gulp.src( "src/lux-autohost.js" )
		.pipe( imports() )
		.pipe( hintNot() )
		.pipe( header( banner, {
			pkg: pkg
		} ) )
		.pipe( rename( "lux-autohost-es6.js" ) )
		.pipe( gulp.dest( "lib/" ) );
} );

function buildEs5() {
	return gulp.src( "src/lux-autohost.js" )
		.pipe( imports() )
		.pipe( hintNot() )
		.pipe( sourcemaps.init() )
		.pipe( babel( {
			compact: false,
			presets: [
				"es2015-without-strict",
				"stage-0"
			]
		} ) )
		.pipe( header( banner, {
			pkg: pkg
		} ) )
		.pipe( sourcemaps.write() )
		.pipe( rename( "lux-autohost.js" ) );
}

gulp.task( "build:es5:quick", function() {
	// build for testing, format to prevent source control showing changes
	return buildEs5()
		.pipe( jscs( {
			configPath: ".jscsrc",
			fix: true
		} ) )
		.pipe( gulp.dest( "lib/" ) );
} );

gulp.task( "build:es5", function() {
	return buildEs5()
		.pipe( gulp.dest( "lib/" ) )
		.pipe( uglify( {
			compress: {
				negate_iife: false
			}
		} ) )
		.pipe( header( banner, {
			pkg: pkg
		} ) )
		.pipe( rename( "lux-autohost.min.js" ) )
		.pipe( gulp.dest( "lib/" ) );
} );

gulp.task( "default", [ "format" ] );

var mocha = require( "gulp-spawn-mocha" );
gulp.task( "test", [ "build:es5:quick" ], function() {
	return gulp.src( [ "spec/**/*.spec.js" ], { read: false } )
		.pipe( mocha( {
			require: [ "spec/helpers/node-setup.js" ],
			reporter: "spec",
			colors: true,
			inlineDiffs: true,
			debug: false
		} ) )
		.on( "error", console.warn.bind( console ) );
} );

gulp.task( "watch", function() {
	gulp.watch( "src/**/*", [ "default" ] );
	gulp.watch( "{lib,spec}/**/*", [ "test" ] );
} );

gulp.task( "format", [ "build:es6", "build:es5" ], function() {
	return gulp.src( [ "**/*.js", "!node_modules/**" ] )
		.pipe( jscs( {
			configPath: ".jscsrc",
			fix: true
		} ) )
		.on( "error", function( error ) {
			gutil.log( gutil.colors.red( error.message ) );
			this.end();
		} )
		.pipe( gulpChanged( ".", { hasChanged: gulpChanged.compareSha1Digest } ) )
		.pipe( gulp.dest( "." ) );
} );


src/batchManager.js
/* global lux, machina, postal, configuration */
/* jshint -W098 */

var metricsClient = {
	action: "sendMetricsBatch",
	queue: [],
	config: configuration.metrics
};
var loggingClient = {
	action: "sendLogBatch",
	queue: [],
	config: configuration.logging
};

var batchManager = new machina.BehavioralFsm( {
	initialState: "queueing",
	states: {
		// Might need an initializing state
		queueing: {
			_onEnter( client ) {
				client.timeout = setTimeout( () => this.handle( client, "transmit" ), client.config.timeout );
			},
			queueEntry( client, data ) {
				client.queue.push( data );
				if ( client.queue.length >= client.config.messages ) {
					this.transition( client, "transmitting" );
				}
			},
			transmit: "transmitting"
		},
		transmitting: {
			_onEnter( client ) {
				clearTimeout( client.timeout );
				this.handle( client, "transmit" );
			},
			transmit( client ) {
				if ( client.queue.length ) {
					var queue = client.queue;
					client.queue = [];
					lux.dispatch( client.action, queue );
				}
				this.transition( client, "queueing" );
			}
		}
	}
} );

var batchListener = lux.actionListener( {
	handlers: {
		sendLogEntry( data ) {
			batchManager.handle( loggingClient, "queueEntry", data );
		},
		sendMetricsEntry( data ) {
			batchManager.handle( metricsClient, "queueEntry", data );
		}
	}
} );


src/actionProcessor.js
/* global moment, postal, configuration, machina, metricsApi, lux */
/* jshint -W098 */

var ignoredActions = [ "sendLogEntry", "sendMetricsEntry", "sendLogBatch", "sendMetricsBatch" ];

function isMonitored( action ) {
	return ignoredActions.indexOf( action ) === -1 &&
		(
			( !configuration.filter.include && configuration.filter.actions.indexOf( action ) === -1 ) ||
			( configuration.filter.include && configuration.filter.actions.indexOf( action ) > -1 )
		);
}

function getContextKey() {
	var args = Array.from( arguments );
	var host;
	var path;
	if ( lux.getContextKey ) {
		return lux.getContextKey( ...arguments );
	} else {
		host = window.location.host.replace( /\./gi, "-" );
		path = window.location.pathname.replace( /\//ig, "-" );
		path = path[ 0 ] === "-" ? path.slice( 1 ) : path;
		return ( [ host, path ].concat( args ) ).join( "." );
	}
}

var actionProcessor = new machina.Fsm( {
	initialize: function() {
		lux.dispatcher.on( "handling", ( data ) => {
			if ( data.inputType === "action.dispatch" ) {
				this.handle( "start", data.client.action.actionType );
			}
		} );
		lux.dispatcher.on( "transition", ( data ) => {
			if ( data.toState === "ready" ) {
				this.handle( "dispatcher.complete" );
			}
		} );
	},
	initialState: "ready",
	states: {
		ready: {
			_onEnter() {
				this.current = {};
			},
			start( action ) {
				if ( isMonitored( action ) ) {
					this.current = {
						name: action,
						start: moment.utc(),
						key: getContextKey( action )
					};
					setTimeout( () => this.handle( "action.complete" ), 0 );
					this.transition( "processing" );
				}
			}
		},
		processing: {
			_onEnter() {
				// ( key, value, unit, customData )
				metricsApi.meter( this.current.key, 1, "count", {} );
			},
			"action.complete": "ready",
			"*": function() {
				this.deferUntilTransition();
			},
			_onExit() {
				// ( key, value, customData )
				metricsApi.timer( this.current.key, moment.utc().diff( this.current.start ), {} );
			}
		}
	}
} );


lib/lux-autohost-es6.js
/**
 * lux-autohost - Action Creator API for submitting logging & metrics to autohost from lux.js apps.
 * Author: Jim Cowart
 * Version: v0.2.0
 * Url: https://github.com/LeanKit-Labs/lux-autohost
 * License(s): MIT Copyright (c) 2016 LeanKit
 */


( function( root, factory ) {
	/* istanbul ignore next - don't test UMD wrapper */
	if ( typeof define === "function" && define.amd ) {
		// AMD. Register as an anonymous module.
		define( [ "lux.js", "postal", "machina", "lodash", "moment" ], factory );
	} else if ( typeof module === "object" && module.exports ) {
		// Node, or CommonJS-Like environments
		module.exports = factory( require( "lux.js" ), require( "postal" ), require( "machina" ), require( "lodash" ), require( "moment" ) );
	} else {
		root.luxah = factory( root.lux, root.postal, root.machina, root._, root.moment );
	}
}( this, function( lux, postal, machina, _, moment ) {
	var configuration = {
		actionChannel: postal.channel( "lux.action" ),
		filter: {
			include: false,
			actions: []
		},
		metrics: {
			timeout: 30000,
			messages: 500
		},
		logging: {
			timeout: 5000,
			messages: 25
		}
	};

	function config( options ) {
		if ( options ) {
			_.merge( configuration, options );
		}
		return configuration;
	}

	lux.customActionCreator( {
		sendLogBatch() {
			var args = Array.from( arguments );
			configuration.actionChannel.publish( {
				topic: "execute.sendLogBatch",
				data: {
					actionType: "sendLogBatch",
					actionArgs: args
				}
			} );
		},
		sendMetricsBatch() {
			var args = Array.from( arguments );
			configuration.actionChannel.publish( {
				topic: "execute.sendMetricsBatch",
				data: {
					actionType: "sendMetricsBatch",
					actionArgs: args
				}
			} );
		}
	} );

	

var logLevels = [ "error", "warn", "info", "debug" ];

function formatLogEntry( type, data ) {
	var msg = data;
	var ns = lux.getLogNamespace ? lux.getLogNamespace( type, data ) : "lux";

	msg = {
		data: data,
		location: window.location.href,
		userAgent: window.navigator.userAgent
	};

	return {
		namespace: ns,
		msg: msg,
		timestamp: moment.utc().toISOString(),
		type: type,
		level: logLevels.indexOf( type ) + 1
	};
}

function logIt( type, data ) {
	lux.dispatch( "sendLogEntry", formatLogEntry( type, data ) );
}

var loggingApi = _.reduce(
	logLevels,
	( acc, level ) => {
		acc[ level ] = function( data ) {
			return logIt( level, data );
		};
		return acc;
	},
	{}
);

lux.customActionCreator( loggingApi );

	


// Need to figure out about custom metadata
function formatMetricsEntry( type, key, value, unit, data ) {
	return {
		type: type, // "time" | "meter" | [custom value]
		key: key, // your metric key
		timestamp: moment.utc().toISOString(),
		value: value,
		units: unit
	};
}

var metricsApi = {
	meter: function( key, value, unit, customData ) {
		lux.dispatch( "sendMetricsEntry", formatMetricsEntry( "meter", key, value, unit, customData ) );
	},
	timer: function( key, value, customData ) {
		lux.dispatch( "sendMetricsEntry", formatMetricsEntry( "timer", key, value, "ms", customData ) );
	}
};

lux.customActionCreator( metricsApi );

	


var metricsClient = {
	action: "sendMetricsBatch",
	queue: [],
	config: configuration.metrics
};
var loggingClient = {
	action: "sendLogBatch",
	queue: [],
	config: configuration.logging
};

var batchManager = new machina.BehavioralFsm( {
	initialState: "queueing",
	states: {
		// Might need an initializing state
		queueing: {
			_onEnter( client ) {
				client.timeout = setTimeout( () => this.handle( client, "transmit" ), client.config.timeout );
			},
			queueEntry( client, data ) {
				client.queue.push( data );
				if ( client.queue.length >= client.config.messages ) {
					this.transition( client, "transmitting" );
				}
			},
			transmit: "transmitting"
		},
		transmitting: {
			_onEnter( client ) {
				clearTimeout( client.timeout );
				this.handle( client, "transmit" );
			},
			transmit( client ) {
				if ( client.queue.length ) {
					var queue = client.queue;
					client.queue = [];
					lux.dispatch( client.action, queue );
				}
				this.transition( client, "queueing" );
			}
		}
	}
} );

var batchListener = lux.actionListener( {
	handlers: {
		sendLogEntry( data ) {
			batchManager.handle( loggingClient, "queueEntry", data );
		},
		sendMetricsEntry( data ) {
			batchManager.handle( metricsClient, "queueEntry", data );
		}
	}
} );

	


var ignoredActions = [ "sendLogEntry", "sendMetricsEntry", "sendLogBatch", "sendMetricsBatch" ];

function isMonitored( action ) {
	return ignoredActions.indexOf( action ) === -1 &&
		(
			( !configuration.filter.include && configuration.filter.actions.indexOf( action ) === -1 ) ||
			( configuration.filter.include && configuration.filter.actions.indexOf( action ) > -1 )
		);
}

function getContextKey() {
	var args = Array.from( arguments );
	var host;
	var path;
	if ( lux.getContextKey ) {
		return lux.getContextKey( ...arguments );
	} else {
		host = window.location.host.replace( /\./gi, "-" );
		path = window.location.pathname.replace( /\//ig, "-" );
		path = path[ 0 ] === "-" ? path.slice( 1 ) : path;
		return ( [ host, path ].concat( args ) ).join( "." );
	}
}

var actionProcessor = new machina.Fsm( {
	initialize: function() {
		lux.dispatcher.on( "handling", ( data ) => {
			if ( data.inputType === "action.dispatch" ) {
				this.handle( "start", data.client.action.actionType );
			}
		} );
		lux.dispatcher.on( "transition", ( data ) => {
			if ( data.toState === "ready" ) {
				this.handle( "dispatcher.complete" );
			}
		} );
	},
	initialState: "ready",
	states: {
		ready: {
			_onEnter() {
				this.current = {};
			},
			start( action ) {
				if ( isMonitored( action ) ) {
					this.current = {
						name: action,
						start: moment.utc(),
						key: getContextKey( action )
					};
					setTimeout( () => this.handle( "action.complete" ), 0 );
					this.transition( "processing" );
				}
			}
		},
		processing: {
			_onEnter() {
				// ( key, value, unit, customData )
				metricsApi.meter( this.current.key, 1, "count", {} );
			},
			"action.complete": "ready",
			"*": function() {
				this.deferUntilTransition();
			},
			_onExit() {
				// ( key, value, customData )
				metricsApi.timer( this.current.key, moment.utc().diff( this.current.start ), {} );
			}
		}
	}
} );


	// jshint ignore: start
	return {
		config,
		actionProcessor,
		batchManager,
		batchListener,
		metricsClient,
		loggingClient
	};
	// jshint ignore: end
} ) );
